### PR TITLE
Implemented VS API for toggling of C++20 modules and STL module build

### LINKED ISSUE #2105: Toggling Visual Studio's new C++ Modules on/off in premake lua script
**What problem will this solve?**
I have an issue with Visual Studio projects using "c++latest" and precompiled headers. Visual Studio has some new options for projects regarding their C++modules which can be found in the project's properties under `C/C++ -> Language`. They are `Enable Experimental C++ Standard Library Modules` and `Build ISO C++23 Standard Library Modules`.

Whenever I make changes to the premake script and regenerate the project these options are set to the standard (enabled) and the code won't compile because of missing include of my precompiled header in their files. I always have to manually set those options again after generating the solution/project.

**What might be a solution?**
Two new commands for the premake lua script that describes the project which can be used to set those values.

```
enablemodules "yes"
buildstlmodules "yes"
```
```
enablemodules "no"
buildstlmodules "no"
```

This way I can set those values already when generating the project via premake.

**What other alternatives have you already considered?**
I don't see any alternatives, but to set those values manually in the project settings.

**Anything else we should know?**
Not really.

### PR DESCRIPTION
**Pull request by nickclark2016**
**What does this PR do?**

Closes #2105 by introducing 2 new APIs.  Introduces `buildstlmodules` and `enablemodules` to the `vsXXXX` exporters.  `buildstlmodules` controls if the STL modules should be built for C++23 and `enablemodules` controls if modules are to be built.  This API does not affect if module files are added to the project or not.

Example Usage:
```lua
buildstlmodules 'Off' -- affects VS2022 only
enablemodules 'On' -- affects VS2019 and VS2022
```

**How does this PR change Premake's behavior?**

Does not affect existing scripts.

**Anything else we should know?**

N/A

**Did you check all the boxes?**

- [x] Focus on a single fix or feature; remove any unrelated formatting or code changes
- [x] Add unit tests showing fix or feature works; all tests pass
- [x] Mention any [related issues](https://github.com/premake/premake-core/issues) (put `closes #XXXX` in comment to auto-close issue when PR is merged)
- [x] Follow our [coding conventions](https://github.com/premake/premake-core/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] Minimize the number of commits
- [x] Align [documentation](https://github.com/premake/premake-core/tree/master/website) to your changes

*You can now [support Premake on our OpenCollective](https://opencollective.com/premake). Your contributions help us spend more time responding to requests like these!*

### CODE REVIEW

File: website/docs/buildstlmodules.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+- `On`
+- `Off`
+
+### Applies To ###
+
+The `config` scope.
+
+### Availability ###
+
+Premake 5.0.0 beta 3 or later.
```
Comment (Jarod42) [inline website/docs/buildstlmodules.md:20]:
```suggestion
Premake 5.0.0 beta 3 or later for Visual Studio 2022 and later.
```

File: website/docs/enablemodules.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+- `On`
+- `Off`
+
+### Applies To ###
+
+The `config` scope.
+
+### Availability ###
+
+Premake 5.0.0 beta 3 or later for Visual Studio 2019 and later.
```
Comment (Jarod42) [inline website/docs/enablemodules.md:20]: How about "See also" section to referencing each other?

File: modules/vstudio/vs2010_vcxproj.lua [hunk 3]
```diff
@@ -3024,6 +3026,24 @@
...
+			if cfg.enablemodules then
+				m.element("EnableModules", nil, iif(cfg.enablemodules == "On", "true", "false"))
+			end
+		end
+	end
+
+
+	function m.buildStlModules(cfg)
+		if _ACTION >= "vs2022" then
+			if cfg.buildstlmodules then
```
[the remark this answers is not in the document]
Comment (nickclark2016) [reply modules/vstudio/vs2010_vcxproj.lua:3040]: In my opinion, it makes no sense to have this activated wtihout the other, but I don't believe we do this kind of sanity checking elsewhere.

[the remark this answers is not in the document]
Comment (Jarod42) [reply modules/vstudio/vs2010_vcxproj.lua:3040]: ok

**Comment by vulcan-dev** [comment]
I appreciate you adding this, makes it easier for a lot of us to experiment and or use it.
However, I do have a question regarding `ScanSourceForModuleDependencies`, I've had to enable that for anything to really work, otherwise I can't import things from the standard library or `Windows.h`. Do you think you could possibly add that if you have free time? If not, then I think I could add it soon because I see how you've added the others.

**Comment by nickclark2016** [comment]
> I appreciate you adding this, makes it easier for a lot of us to experiment and or use it. However, I do have a question regarding `ScanSourceForModuleDependencies`, I've had to enable that for anything to really work, otherwise I can't import things from the standard library or `Windows.h`. Do you think you could possibly add that if you have free time? If not, then I think I could add it soon because I see how you've added the others.

Does `scanformoduledependencies` fit your needs? It is an existing API.

**Comment by nickclark2016** [comment]
> > I appreciate you adding this, makes it easier for a lot of us to experiment and or use it. However, I do have a question regarding `ScanSourceForModuleDependencies`, I've had to enable that for anything to really work, otherwise I can't import things from the standard library or `Windows.h`. Do you think you could possibly add that if you have free time? If not, then I think I could add it soon because I see how you've added the others.
> 
> Does `scanformoduledependencies` fit your needs? It is an existing API.

Bumping this @vulcan-dev

**Comment by vulcan-dev** [comment]
> > > I appreciate you adding this, makes it easier for a lot of us to experiment and or use it. However, I do have a question regarding `ScanSourceForModuleDependencies`, I've had to enable that for anything to really work, otherwise I can't import things from the standard library or `Windows.h`. Do you think you could possibly add that if you have free time? If not, then I think I could add it soon because I see how you've added the others.
> > 
> > Does `scanformoduledependencies` fit your needs? It is an existing API.
> 
> Bumping this @vulcan-dev 

I apologise, I must have had to go when I was responding last time.

It does not, I remember having issues regarding modules, and I had to enable the other option. In my temporary fix, I have it enabled as well and that's what worked for me.

**Comment by nickclark2016** [comment]
> > > > I appreciate you adding this, makes it easier for a lot of us to experiment and or use it. However, I do have a question regarding `ScanSourceForModuleDependencies`, I've had to enable that for anything to really work, otherwise I can't import things from the standard library or `Windows.h`. Do you think you could possibly add that if you have free time? If not, then I think I could add it soon because I see how you've added the others.
> > > 
> > > 
> > > Does `scanformoduledependencies` fit your needs? It is an existing API.
> > 
> > 
> > Bumping this @vulcan-dev
> 
> I apologise, I must have had to go when I was responding last time.
> 
> It does not, I remember having issues regarding modules, and I had to enable the other option. In my temporary fix, I have it enabled as well and that's what worked for me.

So in the fix from this code, the `scanformoduledependencies` does not work? That's weird, since that targets the `ScanSourceForModuleDependencies` property.

**Comment by vulcan-dev** [comment]
> > > > > I appreciate you adding this, makes it easier for a lot of us to experiment and or use it. However, I do have a question regarding `ScanSourceForModuleDependencies`, I've had to enable that for anything to really work, otherwise I can't import things from the standard library or `Windows.h`. Do you think you could possibly add that if you have free time? If not, then I think I could add it soon because I see how you've added the others.
> > > > 
> > > > 
> > > > Does `scanformoduledependencies` fit your needs? It is an existing API.
> > > 
> > > 
> > > Bumping this @vulcan-dev
> > 
> > I apologise, I must have had to go when I was responding last time.
> > 
> > It does not, I remember having issues regarding modules, and I had to enable the other option. In my temporary fix, I have it enabled as well and that's what worked for me.
> 
> So in the fix from this code, the `scanformoduledependencies` does not work? That's weird, since that targets the `ScanSourceForModuleDependencies` property.

If that's the case, then it would work. I thought it was a different option, but I thought wrong. My apologies

**Comment by nickclark2016** [comment]
If that doesn't work when we merge this in, open up a bug and I'll look.

File: modules/vstudio/_preload.lua [hunk 4]
```diff
@@ -195,6 +195,26 @@
...
 	}
 
+	p.api.register {
+		name = "enablemodules",
+		scope = { "config" },
+		kind = "string",
+		allowed = {
+			"On",
+			"Off"
+		}
```
Comment (samsinsane) [inline modules/vstudio/_preload.lua:205]: Using the `boolean` type would probably be better.

File: modules/vstudio/_preload.lua [hunk 5]
```diff
@@ -195,6 +195,26 @@
...
+	}
+
+	p.api.register {
+		name = "buildstlmodules",
+		scope = { "config" },
+		kind = "string",
+		allowed = {
+			"On",
+			"Off"
+		}
```
Comment (samsinsane) [inline modules/vstudio/_preload.lua:215]: Same here.

**Review by samsinsane — APPROVED** [review]

File: modules/vstudio/_preload.lua (hunk 4 above)
Comment (nickclark2016) [reply modules/vstudio/_preload.lua:205]: I thought about using those, but there are also plenty of spots in the code where "Off" and "On" are used in place of a boolean value. Should we start just using boolean instead? If so, that's an easy fix.

Comment (samsinsane) [reply modules/vstudio/_preload.lua:205]: The inputs for boolean are more flexible:
https://github.com/premake/premake-core/blob/561f9a97d3f68fce1f0340bf33213ba8aa588806/src/base/api.lua#L762-L787

From memory there are issues when using this with look up tables, which is how the toolsets work. So, it's up to you whether you think it's a good idea or not.

**Review by Jarod42 — APPROVED** [review]

**Comment by HerrDingenz** [comment]
did this got implemented in a new release already?

**Comment by HerrDingenz** [comment]
nevermind - i just noticed it will be in Premake 5.0.0 beta 3

**Pull request merged**